src/sandstorm/sandstorm-http-bridge.c++

```diff
@@ -456,22 +452,114 @@ private:
   kj::Vector<char> body;
   kj::Vector<Cookie> cookies;
   kj::String statusString;
+  bool headersComplete = false;
+  bool messageComplete = false;
   bool isStreaming = false;
+  bool streamDone = false;
+  bool readStalled = false;
+  bool aborted = false;
+
+  kj::Maybe<kj::Own<kj::PromiseFulfiller<void>>> writeReady;
+  capnp::Request<ByteStream::WriteParams, ByteStream::WriteResults> nextWrite = nullptr;
+  capnp::Orphan<capnp::Data> nextWriteData;
+  size_t nextWriteSize = 0;  // how many bytes are already in `nextWriteData`
+
+  kj::Own<kj::AsyncIoStream> responseInput;
+  byte buffer[8192];
+
+  kj::Promise<void> pumpWrites() {
+    if (nextWriteSize > 0) {
+      // Send the current write and allocate a new one.
+      nextWriteData.truncate(nextWriteSize);
+      nextWrite.adoptData(kj::mv(nextWriteData));
+
+      auto result = nextWrite.send().then([this](auto&&) {
+        return pumpWrites();
+      });
+
+      allocateNextWrite();
+
+      return result;
+    } else if (streamDone) {
+      // No more bytes coming.
+      nextWrite = nullptr;
+      return responseStream.doneRequest().send().ignoreResult();
+    } else {
+      // No bytes received yet. Wait.
+      auto paf = kj::newPromiseAndFulfiller<void>();
+      writeReady = kj::mv(paf.fulfiller);
+      return paf.promise.then([this]() { return pumpWrites(); });
+    }
+  }
+
+  void allocateNextWrite(kj::ArrayPtr<const byte> initData = nullptr) {
+    // For each write we start out allocating twice as much space as we actually managed to fill
+    // on the previous write, though we cap this at 128k.
+    size_t size = nextWriteSize * 2;
+    if (size < sizeof(buffer)) {
+      size = sizeof(buffer);
+    } else if (size > (128u << 10)) {
+      size = (128u << 10);
+    }
+
+    size = kj::max(size, initData.size());
+
+    nextWrite = responseStream.writeRequest();
+    nextWriteData = capnp::Orphanage::getForMessageContaining(
+        ByteStream::WriteParams::Builder(nextWrite))
+        .newOrphan<capnp::Data>(size);
+
+    nextWriteSize = initData.size();
+    if (initData.size() > 0) {
+      memcpy(nextWriteData.get().begin(), initData.begin(), initData.size());
+    }
+
+    if (readStalled) {
+      // Start reading again.
+      readStalled = false;
+      taskSet.add(pumpStreamInternal());
```

Comment:
Ok, so you've got these two concurrent tasks than can mutually wake each other up: `pumpWrites()` and `pumpStreamInternal()`. The latter wakes up the former using the `writeReady()` fulfiller, and former wakes up the latter with this `readStalled` conditional statement here. I think it be somewhat clearer if, instead, there was a `readReady()` fulfiller, so that `pumpStreamInternal()` would only ever be called by itself or by `pumpStream()`.